app/code/Magento/UrlRewrite/Model/Storage/DbStorage.php

```diff
@@ -258,14 +260,17 @@ private function prepareUniqueEntities(array $urls): array
             }
             $uniqueEntities[$url->getStoreId()][$url->getEntityType()] = $entityIds;
         }
+
         return $uniqueEntities;
     }
 
     /**
      * @inheritDoc
      */
-    protected function doReplace(array $urls)
+    protected function doReplace(array $urls): array
     {
+        $this->connection->beginTransaction();
+
         $this->deleteOldUrls($urls);
```

Comment:
if the exception will throw on this level, the transaction will not be rollback properly
please move "try" statement  directly after the transaction begin statement